src/server/builder.ts

```diff
@@ -0,0 +1,417 @@
+﻿/// <reference path="..\compiler\commandLineParser.ts" />
+/// <reference path="..\services\services.ts" />
+/// <reference path="protocol.d.ts" />
+/// <reference path="session.ts" />
+/// <reference types="node" />
+
+namespace ts.server {
+
+    interface Hash {
+        update(data: any, input_encoding?: string): Hash;
+        digest(encoding: string): any;
+    }
+
+    const crypto: {
+        createHash(algorithm: string): Hash
+    } = require("crypto");
+
+    /**
+     * An abstract file info that maintains a shape signature.
+     */
+    export class BuilderFileInfo {
+
+        private lastCheckedShapeSignature: string;
+
+        constructor(
+            public readonly scriptInfo: ScriptInfo,
+            public readonly project: Project
+        ) {
+        }
+
+        public isExternalModule() {
+            const sourceFile = this.getSourceFile();
+            return isExternalModule(sourceFile) || this.containsPurelyAmbientModules(sourceFile);
+        }
+
+        private containsPurelyAmbientModules(sourceFile: SourceFile) {
+            // ambient module declaration will not create externalModuleIndicator
+            // and we only consider files containing purely ambient module declarations
+            // as external module files
+            for (const statement of sourceFile.statements) {
+                if (statement.kind !== SyntaxKind.ModuleDeclaration || (<ModuleDeclaration>statement).name.kind !== SyntaxKind.StringLiteral) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        public computeHash(text: string): string {
+            return crypto.createHash("md5")
+                .update(text)
+                .digest("base64");
+        }
+
+        public getSourceFile(): SourceFile {
+            return this.project.getSourceFile(this.scriptInfo.fileName);
+        }
+
+        public checkIfShapeSignatureChanged() {

```

Comment:
please rename to `updateShapeSignature` and leave a comment about the return value. Name `checkIfShapeSignatureChanged` implies that this method will only read state without updates but it actually changes the state so two subsequent calls to this method might return different results which might be unexpected.